src/main/java/org/folio/circulation/resources/CirculationRulesResource.java

```diff
@@ -89,19 +108,32 @@ private void get(RoutingContext routingContext) {
   @SuppressWarnings("squid:S2147")
   private void put(RoutingContext routingContext) {
     final Clients clients = Clients.create(new WebContext(routingContext), client);
-    CollectionResourceClient loansRulesClient = clients.circulationRulesStorage();
 
-    if (loansRulesClient == null) {
+    if (clients.circulationRulesStorage() == null) {
       internalError(routingContext.response(),
         "Cannot initialise client to storage interface");
       return;
     }
 
+    getExistingPolicyIds(clients)
+      .thenAccept(result -> proceedWithUpdate(result.value(), routingContext, clients));
+  }
+
+  private void proceedWithUpdate(
+    Map<String, Set<String>> existingPoliciesIds, RoutingContext routingContext, Clients clients) {
+
     JsonObject rulesInput;
     try {
       // try to convert, do not safe if conversion fails
       rulesInput = routingContext.getBodyAsJson();
-      Text2Drools.convert(rulesInput.getString("rulesAsText"));
+      Text2Drools.convert(rulesInput.getString("rulesAsText"),
+        (policyType, policies, token) -> {
+          if (!policyExists(existingPoliciesIds, policyType, policies)) {
+             throw new CirculationRulesException(
+               String.format("The policy %s does not exist", policyType),
+               token.getLine(), token.getCharPositionInLine());
+          }
```

Comment:
Maybe this could be extracted to a method (and a reference used) to make the intent clearer